lib/mayaUsd/render/vp2RenderDelegate/material.cpp

```diff
@@ -2349,28 +2436,132 @@ void HdVP2Material::_UpdateShaderInstance(const HdMaterialNetwork& mat)
+                fallbackColor = fallbackIt->second.UncheckedGet<GfVec4f>();
+            }
+            dummyInfo._texture.reset(
+                _GenerateFallbackTexture(textureMgr, "dummy", VtValue(fallbackColor)));
+        }
+    }
+
+    return dummyInfo;
+}
+
+void HdVP2Material::LoadTextures()
+{
+    std::lock_guard<std::mutex> lock(_textureLoadingMutex);
+    for (const auto& task : _textureLoadingTasks) {
+        task.second->Load();
+    }
+}
+
+void HdVP2Material::ClearPendingTasks()
+{
+    // Inform tasks that have not started or finished that this material object
+    // is no longer valid
+    std::lock_guard<std::mutex> lock(_textureLoadingMutex);
+
+    for (auto& task : _textureLoadingTasks) {
+        if (task.second->Terminate()) {
+            // Delete the pointer: we can only do that outside of the object scope
+            delete task.second;
+        }
+    }
+    // Remove the reference of all the tasks
+    _textureLoadingTasks.clear();
+}
+
+void HdVP2Material::_UpdateLoadedTexture(
+    HdSceneDelegate*     sceneDelegate,
+    const std::string&   path,
+    MHWRender::MTexture* texture,
+    bool                 isColorSpaceSRGB,
+    const MFloatArray&   uvScaleOffset)
+{
+    {
+        std::lock_guard<std::mutex> lock(_textureLoadingMutex);
+
+        HdVP2TextureInfo& info = _textureMap[path];
+        info._texture.reset(texture);
+        info._isColorSpaceSRGB = isColorSpaceSRGB;
+        if (uvScaleOffset.length() > 0) {
+            TF_VERIFY(uvScaleOffset.length() == 4);
+            info._stScale.Set(
+                uvScaleOffset[0], uvScaleOffset[1]); // The first 2 elements are the scale
+            info._stOffset.Set(
+                uvScaleOffset[2], uvScaleOffset[3]); // The next two elements are the offset
+        }
 
-    HdVP2TextureInfo& info = _textureMap[path];
-    info._texture.reset(texture);
-    info._isColorSpaceSRGB = isSRGB;
-    if (uvScaleOffset.length() > 0) {
-        TF_VERIFY(uvScaleOffset.length() == 4);
-        info._stScale.Set(uvScaleOffset[0], uvScaleOffset[1]); // The first 2 elements are the scale
-        info._stOffset.Set(
-            uvScaleOffset[2], uvScaleOffset[3]); // The next two elements are the offset
+        // Mark sprim dirty
+        sceneDelegate->GetRenderIndex().GetChangeTracker().MarkSprimDirty(
+            GetId(), HdMaterial::DirtyResource);
+        // Pop the task object from the container, since this method is
+        // called directly from the task object method `loadOnIdle()`,
+        // we do not handle the deletion here, we will let the
+        // function on idle to delete the task object.
+        _textureLoadingTasks.erase(path);
     }
-    return info;
+
+    // TODO: it might triggers many refresh commands if USD has lots of textures
+    //       consider to consolidate/reduce the refresh
+    MGlobal::executeCommandOnIdle("refresh");
```

Comment:
Here we have to think about the user experience. We need the idle priority of the refreshes to be higher than the background loading priority so that we don't have to wait for all the loading to finish before the viewport updates, that's the whole point.

Given that the refresh priority is higher than the background loading priority, asking for refresh after every texture loads means we will refresh once per texture. The time the draw the viewport is probably going to be longer than the time to load one texture (depending on complexity and size), so the refreshes will limit the rate and which we can load textures. We definitely want fewer refreshes.

If the user is interacting with the scene they'll be triggering refreshes of their own, so we don't need to think about that case too much. If the user is not interacting and just watching the viewport, how many refreshes do we need? Maybe 1 per second or something in that range? We can experiment with the rate limit.

A possible implementation would be save the timestamp of the time the last idle refresh was requested, and only allow subsequent refreshes if X amount of time has past since then.

The other nice thing to do would be when all the idle tasks finish always trigger one refresh. That way if the user is waiting to see their scene complete it'll pop in immediately after finishing.

Reply:
Another thing to consider is using `M3dView::scheduleRefreshAllViews` rather than executeCommandOnIdle. I believe scheduleRefreshAllViews has lower overhead than creating the idle command.

Reply:
`M3dView::scheduleRefreshAllViews` indeed improved the UI responsive, I have switched to use `M3dView::scheduleRefreshAllViews`; I also added a timestamp as you suggested, and for the last texture that has been finished loading, a refresh would be triggered regardless of the timestamp:
https://github.com/Autodesk/maya-usd/pull/1980/files#diff-95070490bea9fe284d8a3c42c85ed209a5341007e11b85dac96e316dcec6e269R2624-R2626